Source/Plugins/bsfVulkanRenderAPI/MacOS/BsMacOSRenderWindow.mm

```diff
@@ -0,0 +1,597 @@
+
+		{
+			ScopedSpinLock lock(getCore()->mLock);
+			getCore()->getSyncedProperties().top = props.top;
+			getCore()->getSyncedProperties().left = props.left;
+			getCore()->getSyncedProperties().width = props.width;
+			getCore()->getSyncedProperties().height = props.height;
+		}
+
+		ct::RenderWindowManager::instance().notifySyncDataDirty(getCore().get());
+		_windowMovedOrResized();
+	}
+
+	void MacOSRenderWindow::setWindowed(UINT32 width, UINT32 height)
+	{
+		RenderWindowProperties& props = mProperties;
+
+		if (!props.isFullScreen)
+			return;
+
+		// Restore original display mode
+		const VideoModeInfo& videoModeInfo = ct::RenderAPI::instance().getVideoModeInfo();
+
+		UINT32 outputIdx = 0; // 0 is always primary
+		if(outputIdx >= videoModeInfo.getNumOutputs())
+		{
+			LOGERR("Invalid output device index.")
+			return;
+		}
+
+		const VideoOutputInfo& outputInfo = videoModeInfo.getOutputInfo(outputIdx);
+		setDisplayMode(outputInfo, outputInfo.getDesktopVideoMode());
+
+		mWindow->setWindowed();
+
+		props.isFullScreen = false;
+		props.width = width;
+		props.height = height;
+
+		{
+			ScopedSpinLock lock(getCore()->mLock);
+			getCore()->getSyncedProperties().width = props.width;
+			getCore()->getSyncedProperties().height = props.height;
+		}
+
+		ct::RenderWindowManager::instance().notifySyncDataDirty(getCore().get());
+		_windowMovedOrResized();
+	}
+
+	void MacOSRenderWindow::setDisplayMode(const VideoOutputInfo& output, const VideoMode& mode)
+	{
+		CGDisplayFadeReservationToken fadeToken = kCGDisplayFadeReservationInvalidToken;
+		if (CGAcquireDisplayFadeReservation(5.0f, &fadeToken))
+			CGDisplayFade(fadeToken, 0.3f, kCGDisplayBlendNormal, kCGDisplayBlendSolidColor, 0, 0, 0, TRUE);
+
+		auto& destOutput = static_cast<const ct::MacOSVideoOutputInfo&>(output);
+		auto& newMode = static_cast<const ct::MacOSVideoMode&>(mode);
+
+		// Note: An alternative to changing display resolution would be to only change the back-buffer size. But that doesn't
+		// account for refresh rate, so it's questionable how useful it would be.
+		CGDirectDisplayID displayID = destOutput._getDisplayID();
+		CGDisplaySetDisplayMode(displayID, newMode._getModeRef(), nullptr);
+
+		if (fadeToken != kCGDisplayFadeReservationInvalidToken)
+		{
+			CGDisplayFade(fadeToken, 0.3f, kCGDisplayBlendSolidColor, kCGDisplayBlendNormal, 0, 0, 0, FALSE);
+			CGReleaseDisplayFadeReservation(fadeToken);
+		}
+	}
+
+	Vector2I MacOSRenderWindow::screenToWindowPos(const Vector2I& screenPos) const
+	{
+		return mWindow->screenToWindowPos(screenPos);
+	}
+
+	Vector2I MacOSRenderWindow::windowToScreenPos(const Vector2I& windowPos) const
+	{
+		return mWindow->windowToScreenPos(windowPos);
+	}
+
+	SPtr<ct::MacOSRenderWindow> MacOSRenderWindow::getCore() const
+	{
+		return std::static_pointer_cast<ct::MacOSRenderWindow>(mCoreSpecific);
+	}
+    
+    void MacOSRenderWindow::_windowMovedOrResized()
+    {
+        THROW_IF_NOT_CORE_THREAD;
```

Comment:
This seems to have a lot less code than the OpenGL version, any reason?

Reply:
i redirected it to ct::BsMacOSRenderWindow

Reply:
This method will get called on the main thread, while everything in the ct:: namespace executes on the core thread. It's not safe to call core thread methods directly, plus its method does different things like rebuilding the swap chain, while this method should just update some local values. Also `THROW_IF_NOT_CORE_THREAD` should trigger here, as this method shouldn't be called on the core thread unless I'm misremembering something.

Reply:
ok, this is a very important thing, thanks ;)